core/state.py

```diff
@@ -2297,6 +2297,8 @@ def Unset(self, lval, which_scopes):
                     # Special case: The array SHORTENS if you unset from the end.  You
                     # can tell with a+=(3 4)
```

Comment:
Hm I don't recall this behavior

What does bash do if you `unset a[-2]` or `a[-3]`?    Is it documented anywhere?

Reply:
I wonder if we have test coverage of `a[-2]` and a[-len]` and so forth??  If so maybe it is a good idea to add it?

Reply:
> What does bash do if you `unset a[-2]` or `a[-3]`? Is it documented anywhere?

The problem that this PR tries to fix is unrelated to the behavior for `unset a[-2]` in the present case. Anyway, `a[-2]` and `a[-3}` removes an element `a[n-2]` and `a[n-3]`, respectively, with `n` being the index of the maximum set element plus one.

The present problem is that after `a[9]=x; unset -v "a[-1]"`, the array does not become the original state in the current `master` of `osh`. With `a[9]`, the array becomes `(1 None None None None None None None None x)`. Then, with `unset -v "a[-1]"`, the array should become `(1)`. However, in the current `master` of `osh`, it results in `(1 None None None None None None None None)`. Then, after that, `${a[@]: -1}` in `osh` picks up the last element of the filler `None`. This doesn't make sense.

Reply:
> I wonder if we have test coverage of `a[-2]` and `a[-3]` and `a[-len]` and so forth?? If so maybe it is a good idea to add it?

I think that is covered by the test added by #2153.

Reply:
Maybe I have confused you by showing the comparison with Bash. Even if we do not compare the behavior with Bash, the current `osh` behavior is absolutely strange.

```bash
a[0]=x
a[2]=y
unset -v 'a[2]'
# Here, we expect the state to be equivalent to a=(x)
echo "${a[@]: -1}"  # --> This produces no words.
```